src/Measurement_Observation/haskell_model.rst

```diff
@@ -0,0 +1,129 @@
+
+Haskell Compilation and Execution Model
+=======================================
+
+
+GHC doesn't use a VM but compiles to native code (when a native code generator
+is used).
+So we can use tools to profile native code execution: e.g. perf.
+However the native code we execute doesn't look exactly like the one produced by
+imperative languages (C, C++, Rust, etc.) and this confuses the tools which make
+some assumptions.
+For example:
+
+1. we don't have clear procedure boundaries with call/return instruction pairs
+   (we use tail calls, i.e. jump instructions)
+
+2. we don't use the so-called C stack in the usual way. I.e. we don't use usual
+   stack registers (e.g. rsp/rbp on x86-64) as stack registers.
+
+3. GHC uses its own calling convention
+
+In addition, lazy evaluation makes control-flow and memory usage difficult to
+understand.
+
+In this chapter we describe the compilation pipeline and the execution model.
+For each stage of the pipeline we list what can be measured at runtime.
+
+
+Compilation pipeline overview
+-----------------------------
+
+GHC compiles Haskell codes in several stages, each with their own intermediate
+representation:
+
+Haskell: the functional language we love with its bazillion of extensions.
+
+Core: a simple and explicitly typed functional language. Types are basically
+Haskell types. Type-class dictionaries are passed explicitly as records,
+type-applications are used everywhere needed, coercions (proofs that we can
+convert a type into another) are first class values, etc.
+
+STG: a functional language closer to the execution model. It's still typed but
+with primitive types. E.g. it tracks if a value is a heap object or an
+unboxed word (unboxed = doesn't have its own heap object) but not the Haskell
+type of the heap object. Complex forms are lowered to simpler ones (e.g. unboxed
+sums are lowered into unboxed tuples); this is called unarisation.
+
+Additionally every heap allocation is made explicit with a let-binding. For
+example instead of ``foo (MkBar x y)`` you have ``let b = MkBar x y in foo b``.
+This is called `A-normal form<https://en.wikipedia.org/wiki/A-normal_form>`_.
+During code generation from STG we know that all functions are applied to
+"simple" arguments only (variables, constants, etc.).
```

Comment:
```suggestion
STG comports to `A-normal form<https://en.wikipedia.org/wiki/A-normal_form>`_ by upholding two invariants. First, every heap allocation is made explicit with a let-binding; for
example instead of ``foo (MkBar x y)`` in STG we have ``let b = MkBar x y in foo b``. Second, all functions are applied to
"simple" arguments only (variables, constants, etc.). 

These features make STG a good intermediate representation to read to understand exactly where memory is being allocated. By the time the code is transformed into STG all polymorphism and type level features have been compiled away, simplifying the program. Because STG is in a normal form, one can be confident that every let observed in the STG representation of their program is doing allocation; which is not true for Haskell or Core.
```